system-helper/flatpak-system-helper.c

```diff
@@ -141,6 +206,137 @@ no_progress_cb (OstreeAsyncProgress *progress, gpointer user_data)
 {
 }
 
+static uid_t
+get_connection_uid (GDBusMethodInvocation *invocation, GError **error)
+{
+  GDBusConnection *connection = g_dbus_method_invocation_get_connection (invocation);
+  const gchar *sender = g_dbus_method_invocation_get_sender (invocation);
+  g_autoptr(GVariant) credentials = NULL;
+  g_autoptr(GVariantIter) iter = NULL;
+  const char *key = NULL;
+  g_autoptr(GVariant) value = NULL;
+  uid_t uid;
+
+  credentials = g_dbus_connection_call_sync (connection,
+                                             "org.freedesktop.DBus",
+                                             "/org/freedesktop/DBus",
+                                             "org.freedesktop.DBus",
+                                             "GetConnectionCredentials",
+                                             g_variant_new ("(s)", sender),
+                                             NULL, G_DBUS_CALL_FLAGS_NONE, G_MAXINT, NULL, error);
+
+  if (credentials == NULL)
+    {
+      g_dbus_method_invocation_return_error (invocation, G_DBUS_ERROR, G_DBUS_ERROR_FAILED,
+                                             "Error calling GetConnectionCredentials over D-Bus");
+      return TRUE;
+    }
+
+  g_variant_get (credentials, "(a{sv})", &iter);
```

Comment:
Assuming the function has `gboolean` return type and a `uid_t *uid_out` parameter, this loop can be replaced with something like:

```
g_autoptr(GVariant) dict = g_variant_get_child_value (credentials, 0);
g_autoptr(GVariant) unixUserId = g_variant_lookup_value (dict, "UnixUserID", G_VARIANT_TYPE_UINT32);

if (unixUserId == NULL) { set error; return FALSE; }
*uid_out = g_variant_get_uint32 (unixUserId);
return TRUE;
```